### PR TITLE
First version of the Remoting Controller

### PR DESCRIPTION
**Pull request by giftkugel**
Rewritten version of the Polymer Component.

### CODE REVIEW

File: remoting-controller.html [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+                 * @param {any} changes The changes made to the model
+                 */
+                 _modelChanged(changes) {
+
+                    if (changes) {
+                        const arrayChange = changes.path.indexOf('.splices') >= 0;
+                        RemotingController.LOGGER.debug(this.name, 'Model changed', 'Changes:', changes, 'Is Array change:', arrayChange);
+
+                        const { value, base, path } = changes;
+                        const elements = Polymer.Path.split(path);
```
Comment (hastebrot) [inline remoting-controller.html]: Maybe rename to `pathSegments`, because `elements` can also be HTML elements?

File: remoting-controller.html [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+                 *
+                 * @param {object} controllerProxy The `ControllerProxy` from the Dolphin Platform.
+                 */
+                _handleControllerCreated(controllerProxy) {
+                    RemotingController.LOGGER.debug(this.name, 'Dolphin Platform controller created', controllerProxy.getId(), controllerProxy);
+                    controllerProxy.onDestroyed(this._handleControllerDestroyed);
+                    controller = controllerProxy;
+                    let internalModel = controllerProxy.getModel();
+                    this.set('model', internalModel);
+                    calulateBeanPathForModel(internalModel);
```
Comment (hastebrot) [inline remoting-controller.html]: Typo: `calculateBeanPathForModel()`.

File: remoting-controller.html [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+                baseArray.splice.apply(baseArray, [startIndex, 0].concat(insertArray));
+            };
+
+            const beanToPath = new Map();
+
+            const calculateBeanPath = function(path, bean) {
+                if (!exists(bean)) {
+                    return;
+                }
+                beanToPath.set(bean, path);
```
Comment (hastebrot) [inline remoting-controller.html]: `beanToPath` is set to bean, but only cleaned when controller is created or array was spliced. This result in a memory leak.

File: remoting-controller.html (hunk 2 above)
Comment (giftkugel) [reply remoting-controller.html]: fixed.

File: remoting-controller.html (hunk 1 above)
Comment (giftkugel) [reply remoting-controller.html]: ok

File: remoting-controller.html [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+                 * Creates a Dolphin controller and registers the listeners for bean updates.
+                 */
+                _createController() {
+                    this._createMethodObserver('_modelChanged(model.*)', false);
+                    RemotingController.LOGGER.info('Creating Dolphin Platform controller', this.name);
+
+                    // Register callbacks for the Dolphin Platform bean manager
+                    internalClientContext.beanManager.onAdded(this._onAdded);
+                    internalClientContext.beanManager.onRemoved(this._onRemoved);
+                    internalClientContext.beanManager.onBeanUpdate(this._onBeanUpdate);
```
[the remark this answers is not in the document]
Comment (giftkugel) [reply remoting-controller.html]: Does this question belong to line 477, or the block 475-478?
The RemotingController adds his own method as listener on the BeanManager.

[the remark this answers is not in the document]
Comment (giftkugel) [reply remoting-controller.html]: It is called per RemotingController.

File: remoting-controller.html (hunk 3 above)
Comment (giftkugel) [reply remoting-controller.html]: Changes to the model structure are done at list/array level. Once the initialization is done, the model structure does not change anymore. Only lists/arrays can be changed due to add/remove or splice. Those changes are handled as you already mentioned.

Comment (giftkugel) [reply remoting-controller.html]: Right. Working on it.

File: remoting-controller.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (giftkugel) [reply remoting-controller.html]: I moved the clientContext into the controller

File: remoting-controller.html (hunk 3 above)
Comment (giftkugel) [reply remoting-controller.html]: Fixed in the new version. The bean Map will be recreated on bean updates.

**Pull request merged**